File: Python/0101-0150/143. Reorder List.py
# https://leetcode.com/problems/reorder-list/

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # Function to find the mid of the linkedList
        def findMid (head):
            fast, slow = head, head

            while fast != None and fast.next != None:
                slow = slow.next
                fast = fast.next.next

            return slow

        # Function to reverse LinkedList
        def reverseList (Mid):
            current, previous, next = Mid, None, Mid.next

            while current != None:
                current.next = previous
                previous = current
                current = next
                next = next.next if (next!=None) else None

            return previous

        if head == None or head.next == None:
            return

        # Reverse from the Mid of the List
        start, tail = head, reverseList(findMid(head))

        startNext, tailNext = None, None
        while start != None and tail != None:

            startNext = start.next
            start.next = tail

            tailNext = tail.next
            tail.next = None if (startNext == tail) else startNext

            start = startNext
            tail = tailNext
            

File: Javascript/0101-0150/143. Reorder List.js
// https://leetcode.com/problems/reorder-list/

var reorderList = function(head) {

    // Function to find the mid of the linkedList
    var findMid = function(head){
        let fast = head, slow = head;

        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    // Function to reverse LinkedList
    var reverseList = function (Mid){
        let current = Mid, previous = null, next = current.next;

        while(current != null){
            current.next = previous;
            previous = current;
            current = next;
            next = (next!=null) ? next.next : null;
        }

        return previous;
    }

    if(head == null || head.next == null) return;

    // Reverse from the Mid of the List
    let start = head, tail = reverseList(findMid(head));

    let startNext = null, tailNext = null;
    while(start != null && tail != null){

        startNext = start.next;
        start.next = tail;

        tailNext = tail.next;
        tail.next = (startNext == tail) ? null : startNext;

        start = startNext; tail = tailNext;
    }

};
